subprojects/gst-plugins-bad/sys/msdk/gstmsdkdec.c

```diff
@@ -161,13 +161,43 @@ gst_msdkdec_free_unlocked_msdk_surfaces (GstMsdkDec * thiz)
   for (l = thiz->locked_msdk_surfaces; l;) {
     GList *next = l->next;
     surface = l->data;
-    if (surface->surface->Data.Locked == 0) {
+    if (surface->surface->Data.Locked == 0 &&
+        GST_MINI_OBJECT_REFCOUNT (surface->buf) == 1) {
       free_surface (surface);
       thiz->locked_msdk_surfaces =
           g_list_delete_link (thiz->locked_msdk_surfaces, l);
     }
     l = next;
   }
+  /* We need to check if all surfaces are in used */
+  if (g_list_length (thiz->locked_msdk_surfaces) ==
+      thiz->alloc_resp.NumFrameActual)
+    return FALSE;
+  else
+    return TRUE;
+}
+
+static gboolean
+check_surface_available (GstMsdkDec * thiz)
+{
+  GList *l;
+  GstMsdkSurface *surface = NULL;
+  gboolean ret = FALSE;
+
+  for (l = thiz->locked_msdk_surfaces; l;) {
+    GList *next = l->next;
+    surface = l->data;
+    if (surface->surface->Data.Locked == 0
+        && GST_MINI_OBJECT_REFCOUNT (surface->buf) == 1) {
```

Comment:
It's so tricky here. Please at lease use `GST_MINI_OBJECT_REFCOUNT_VALUE`, which is a atomic operation.

Reply:
> It's so tricky here. Please at lease use `GST_MINI_OBJECT_REFCOUNT_VALUE`, which is a atomic operation.

I can not think of another way to check if downstream has released this buffer before we let msdkdec unref it. Do you have a better solution?